networkit/cpp/scd/LocalTightnessExpansion.cpp

```diff
@@ -0,0 +1,297 @@
+        } else {
+            wd += g.degree(u);
+        }
+
+        weightedDegree.push_back(std::sqrt(wd));
+
+        nodeInternalSimilarity.push_back(.0);
+        nodeExternalSimilarity.push_back(.0);
+        inResult.push_back(false);
+        inShell.push_back(false);
+    };
+
+    LocalGraph<is_weighted, decltype(addNode)> localGraph(g, addNode);
+
+    auto updateShell = [&](node u, node lu) {
+#ifndef NDEBUG
+        std::unordered_map<node, double> uNeighbors;
+        g.forNeighborsOf(
+            u, [&](node, node v, edgeweight ew) { uNeighbors.insert(std::make_pair(v, ew)); });
+
+        if (inShell[lu]) {
+            double debugUExternalSimilarity = .0;
+            double debugUInternalSimilarity = .0;
+            localGraph.forNeighborsWithTrianglesOf(u, [&](node lv, edgeweight, double triangleSum) {
+                double scoreUv = 0.0;
+                double denom = weightedDegree[lv] * weightedDegree[lu];
+                if (denom > 0.0) {
+                    scoreUv = triangleSum / denom;
+                }
+
+                if (inResult[lv]) {
+                    debugUInternalSimilarity += scoreUv;
+                } else {
+                    debugUExternalSimilarity += scoreUv;
+                }
+            });
+
+            assert(std::abs(debugUExternalSimilarity - nodeExternalSimilarity[lu]) < 0.000001);
+            assert(std::abs(debugUInternalSimilarity - nodeInternalSimilarity[lu]) < 0.000001);
+        }
+#endif
+
+        std::vector<node> newShellNodes;
+
+        localGraph.forNeighborsWithTrianglesOf(
+            u, [&](node lv, edgeweight weight, double triangleSum) {
+                // collect counts and compute scores
+                double denom = weightedDegree[lv] * weightedDegree[lu];
+                double scoreUv = triangleSum / denom;
+
+#ifndef NDEBUG
+                {
+                    double debugScore = weightedEdgeScore(g, u, localGraph.toGlobal(lv), weight,
+                                                          weightedDegree[lu], uNeighbors);
+                    assert(scoreUv == debugScore);
+                }
+#else
+            tlx::unused(weight); // needed for assert above
+#endif
+
+                nodeInternalSimilarity[lv] += scoreUv;
+
+                if (inResult[lv]) {
+                    externalSimilarity -= scoreUv;
+                    internalSimilarity += 2 * scoreUv;
+                    if (!inShell[lu]) {
+                        nodeInternalSimilarity[lu] += scoreUv;
+                    }
+                    nodeExternalSimilarity[lv] -= scoreUv;
+                } else {
+                    externalSimilarity += scoreUv;
+
+                    if (!inShell[lu]) {
+                        nodeExternalSimilarity[lu] += scoreUv;
+                    }
+
+                    if (shell.contains(lv)) {
+                        shell.update(lv);
+                    } else {
+                        shell.push(lv);
+                    }
```

Comment:
You can simplify this code by simply using `shell.update(lv)`, `update` already implements this check.